rust/feature-flags/src/cohort_operations.rs

```diff
@@ -0,0 +1,351 @@
+    #[instrument(skip_all)]
+    pub async fn from_pg(
+        client: Arc<dyn DatabaseClient + Send + Sync>,
+        cohort_id: i32,
+        team_id: i32,
+    ) -> Result<Cohort, FlagError> {
+        let mut conn = client.get_connection().await.map_err(|e| {
+            tracing::error!("Failed to get database connection: {}", e);
+            // TODO should I model my errors more generally?  Like, yes, everything behind this API is technically a FlagError,
+            // but I'm not sure if accessing Cohort definitions should be a FlagError (vs idk, a CohortError?  A more general API error?)
+            FlagError::DatabaseUnavailable
+        })?;
+
+        let query = "SELECT id, name, description, team_id, deleted, filters, query, version, pending_version, count, is_calculating, is_static, errors_calculating, groups, created_by_id FROM posthog_cohort WHERE id = $1 AND team_id = $2";
+        let cohort_row = sqlx::query_as::<_, CohortRow>(query)
+            .bind(cohort_id)
+            .bind(team_id)
+            .fetch_optional(&mut *conn)
+            .await
+            .map_err(|e| {
+                tracing::error!("Failed to fetch cohort from database: {}", e);
+                FlagError::Internal(format!("Database query error: {}", e))
+            })?;
+
+        match cohort_row {
+            Some(row) => Ok(Cohort {
+                id: row.id,
+                name: row.name,
+                description: row.description,
+                team_id: row.team_id,
+                deleted: row.deleted,
+                filters: row.filters,
+                query: row.query,
+                version: row.version,
+                pending_version: row.pending_version,
+                count: row.count,
+                is_calculating: row.is_calculating,
+                is_static: row.is_static,
+                errors_calculating: row.errors_calculating,
+                groups: row.groups,
+                created_by_id: row.created_by_id,
+            }),
+            None => Err(FlagError::DatabaseError(format!(
+                "Cohort with id {} not found for team {}",
+                cohort_id, team_id
+            ))),
+        }
+    }
+
+    /// Parses the filters JSON into a CohortProperty structure
+    // TODO: this doesn't handle the deprecated "groups" field, see
+    // https://github.com/PostHog/posthog/blob/feat/dynamic-cohorts-rust/posthog/models/cohort/cohort.py#L114-L169
+    // I'll handle that in a separate PR.
+    pub fn parse_filters(&self) -> Result<Vec<PropertyFilter>, FlagError> {
+        let wrapper: serde_json::Value = serde_json::from_value(self.filters.clone())?;
+        let cohort_property: InnerCohortProperty =
+            serde_json::from_value(wrapper["properties"].clone())?;
+        Ok(cohort_property.to_property_filters())
+    }
+}
+
+impl InnerCohortProperty {
+    pub fn to_property_filters(&self) -> Vec<PropertyFilter> {
```

Comment:
generally, `to_` functions consume the `self`. I'd call this just `property_filters`, or even just `filters` (the "property" part is already in the type name).

If you wanted, I've walked the code path and I'm fairly sure this could return an `impl Iterator<Item = PropertyFilter>`, too, but it requires a little bit of doing:
```diff
diff --git a/rust/feature-flags/src/cohort_operations.rs b/rust/feature-flags/src/cohort_operations.rs
index 2839432e04..2b97ac261e 100644
--- a/rust/feature-flags/src/cohort_operations.rs
+++ b/rust/feature-flags/src/cohort_operations.rs
@@ -60,7 +60,7 @@ impl Cohort {
     // TODO: this doesn't handle the deprecated "groups" field, see
     // https://github.com/PostHog/posthog/blob/feat/dynamic-cohorts-rust/posthog/models/cohort/cohort.py#L114-L169
     // I'll handle that in a separate PR.
-    pub fn parse_filters(&self) -> Result<Vec<PropertyFilter>, FlagError> {
+    pub fn parse_filters(&self) -> Result<impl Iterator<Item = PropertyFilter>, FlagError> {
         let wrapper: serde_json::Value = serde_json::from_value(self.filters.clone())?;
         let cohort_property: InnerCohortProperty =
             serde_json::from_value(wrapper["properties"].clone())?;
@@ -69,12 +69,8 @@ impl Cohort {
 }
 
 impl InnerCohortProperty {
-    pub fn to_property_filters(&self) -> Vec<PropertyFilter> {
-        self.values
-            .iter()
-            .flat_map(|value| &value.values)
-            .cloned()
-            .collect()
+    pub fn to_property_filters(self) -> impl Iterator<Item = PropertyFilter> {
+        self.values.into_iter().flat_map(|value| value.values)
     }
 }

diff --git a/rust/feature-flags/src/flag_matching.rs b/rust/feature-flags/src/flag_matching.rs
index 5d40fe951b..e27f24abdf 100644
--- a/rust/feature-flags/src/flag_matching.rs
+++ b/rust/feature-flags/src/flag_matching.rs
@@ -749,7 +749,7 @@ impl FeatureFlagMatcher {
             // Evaluate cohort conditions
             if !cohort_filters.is_empty()
                 && !self
-                    .evaluate_cohort_filters(&cohort_filters, &properties_to_check)
+                    .evaluate_cohort_filters(cohort_filters.into_iter(), &properties_to_check)
                     .await?
             {
                 return Ok((false, FeatureFlagMatchReason::NoConditionMatch));
@@ -824,19 +824,25 @@ impl FeatureFlagMatcher {
         }
     }
 
-    pub async fn evaluate_cohort_filters(
+    pub async fn evaluate_cohort_filters<F>(
         &self,
-        filters: &[PropertyFilter],
+        filters: F,
         target_properties: &HashMap<String, Value>,
-    ) -> Result<bool, FlagError> {
+    ) -> Result<bool, FlagError>
+    where
+        F: Iterator<Item = PropertyFilter>,
+    {
         Box::pin(self.evaluate_potentially_nested_cohort_filters(filters, target_properties)).await
     }
 
-    async fn evaluate_potentially_nested_cohort_filters(
+    async fn evaluate_potentially_nested_cohort_filters<F>(
         &self,
-        filters: &[PropertyFilter],
+        filters: F,
         target_properties: &HashMap<String, Value>,
-    ) -> Result<bool, FlagError> {
+    ) -> Result<bool, FlagError>
+    where
+        F: Iterator<Item = PropertyFilter>,
+    {
         let mut cohort_filters = Vec::new();
         let mut non_cohort_filters = Vec::new();
```

Plus a few other minor fixups.